core/src/main/scala/cats/data/Chain.scala

```diff
@@ -145,6 +145,30 @@ sealed abstract class Chain[+A] {
       else acc
     }
 
+  /**
+   * Finds the first element of this `Chain` for which the given partial
+   * function is defined, and applies the partial function to it.
+   */
+  final def collectFirst[B](pf: PartialFunction[A, B]): Option[B] = {
+    var result: Option[B] = None
+    foreachUntil { a =>
+      // trick from TraversableOnce, used to avoid calling both isDefined and apply (or calling lift)
+      val x = pf.applyOrElse(a, sentinel)
+      if (x.asInstanceOf[AnyRef] ne sentinel) {
+        result = Some(x.asInstanceOf[B])
+        true
+      } else false
+    }
+    result
+  }
+
+  /**
+   * Like `collectFirst` from `scala.collection.Traversable` but takes `A => Option[B]`
+   * instead of `PartialFunction`s.
+   */
+  final def collectFirstSome[B](f: A => Option[B]): Option[B] =
+    collectFirst(Function.unlift(f))
```

Comment:
Can we reimplement using `foreachUntil`? The whole purpose of overriding the Foldable one is the performance gain, so I'd rather be sure that it reduces overhead to the minimum.

Reply:
I'm not 100% sure that this will give some noticeable performance improvement. I will try to benchmark this.

Reply:
Thanks I am curious to see the benchmark `foreachUntil` implementation of `collectFrist` v.s. the vanilla implementation from `Foldable` using `foldRight`

Reply:
I finally find time for the benchmark.

I made a [separate branch](https://github.com/LMnet/cats/tree/collectFirstBench) for benchmarking. [Here](https://github.com/LMnet/cats/commit/9dd3b61620faefec451a3d936044949581433f91) is the commit with the code of the benchmark. 

I removed overrides from the `Foldable` instance to compare the performance of the default implementation of `collectFirst`/`collectFirstSome` from the `Foldable`. Also, I added `collectFirstSome2` with the specialized version of `collectFirstSome`, based on `foreachUntil` and without `Function.unlift`.

I benchmarked small (size 5), medium (size 100) and large (size 1000000) chains.

I ran benchmarks with the following command: `jmh:run -i 10 -wi 10 -f1 -t1 .*ChainCollectFirst.*`

## Benchmarks

```
Benchmark                                               Mode  Cnt         Score       Error  Units
ChainCollectFirstBench.collectFirstSmallFoldable       thrpt   10   7299195.442 ± 55930.435  ops/s
ChainCollectFirstBench.collectFirstSmallNew            thrpt   10  19402233.046 ± 77203.919  ops/s

ChainCollectFirstBench.collectFirstMediumFoldable      thrpt   10    349714.002 ±  1802.971  ops/s
ChainCollectFirstBench.collectFirstMediumNew           thrpt   10   1065836.106 ± 11337.181  ops/s

ChainCollectFirstBench.collectFirstLargeFoldable       thrpt   10    377742.008 ±  1213.396  ops/s
ChainCollectFirstBench.collectFirstLargeNew            thrpt   10   1041808.061 ±  8960.744  ops/s

ChainCollectFirstBench.collectFirstSomeSmallFoldable   thrpt   10   7260583.162 ± 66323.933  ops/s
ChainCollectFirstBench.collectFirstSomeSmallNew        thrpt   10  15223500.005 ± 70156.918  ops/s
ChainCollectFirstBench.collectFirstSome2SmallNew       thrpt   10  23795542.657 ± 53002.514  ops/s

ChainCollectFirstBench.collectFirstSomeMediumFoldable  thrpt   10    376093.351 ±  2306.020  ops/s
ChainCollectFirstBench.collectFirstSomeMediumNew       thrpt   10   1250584.212 ±  8875.084  ops/s
ChainCollectFirstBench.collectFirstSome2MediumNew      thrpt   10   1412017.858 ± 13673.273  ops/s

ChainCollectFirstBench.collectFirstSomeLargeFoldable   thrpt   10    392443.540 ±  2350.104  ops/s
ChainCollectFirstBench.collectFirstSomeLargeNew        thrpt   10   1243549.960 ± 13598.336  ops/s
ChainCollectFirstBench.collectFirstSome2LargeNew       thrpt   10   1399553.286 ±  7312.677  ops/s
```
- `*Foldable` is the default implementation from `Foldable`.
- `*New` is new implementation from my PR
- `collectFirstSome2*New` is `collectFirstSome` based on `foreachUntil` and without `Function.unlift`

## Results

1. Implementations of `collectFirst`/`collectFirstSome` from my PR is few times faster than defaults for any `Chain` size.
2. `collectFirstSome` based on `collectFirst` and `Function.unlift` is ~2 times slower than `collectFirstSome` based on `foreachUntil` for the small chains.
3. `collectFirstSome` based on `collectFirst` and `Function.unlift` has almost same performance as `collectFirstSome` based on `foreachUntil` for the medium and large chains.

## Conclusion

I changed `collectFirstSome` implementation to `foreachUntil`. I added a new commit to this PR.

Reply:
awesome! thanks so much for this contribution.